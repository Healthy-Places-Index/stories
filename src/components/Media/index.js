/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useQuery, useMutation, gql } from '@apollo/client';
import { Segment, Form, Input } from 'semantic-ui-react';

import debouncedMutation from '../../providers/debouncedMutation';

export const GET_SLIDE_MEDIA = gql`
  query GetSlideMedia($slide: ID!) {
    Slide(where: { id: $slide }) {
      id
      youtube
      soundcloud
    }
  }
`;

const UPDATE_SLIDE_YOUTUBE = gql`
  mutation UpdateSlideYoutube($slide: ID!, $youtube: String) {
    updateSlide(id: $slide, data: { youtube: $youtube }) {
      id
      youtube
    }
  }
`;

const UPDATE_SLIDE_SOUNDCLOUD = gql`
  mutation UpdateSlideSoundcloud($slide: ID!, $soundcloud: String) {
    updateSlide(id: $slide, data: { soundcloud: $soundcloud }) {
      id
      soundcloud
    }
  }
`;

const Media = ({ slide }) => {
  const { loading, data } = useQuery(GET_SLIDE_MEDIA, {
    variables: { slide },
  });

  const [youtube, setYoutube] = useState('');
  const [soundcloud, setSoundcloud] = useState('');

  useEffect(() => {
    setYoutube(loading && !data ? '' : data.Slide.youtube || '');
    setSoundcloud(loading && !data ? '' : data.Slide.soundcloud || '');
  }, [loading, data]);

  const [updateYoutube] = useMutation(UPDATE_SLIDE_YOUTUBE);
  const [updateSoundcloud] = useMutation(UPDATE_SLIDE_SOUNDCLOUD);

  const youtubeTimer = useRef();
  const soundcloudTimer = useRef();

  return (
    <Form.Field>
      <label>Media</label>
      <Segment style={{ marginTop: 0 }}>
        <Form.Field>
          <label>Youtube: </label>
          <Input
            placeholder="Youtube URL"
            icon="linkify"
            iconPosition="left"
            value={youtube}
            onChange={(e, { value }) => {
              setYoutube(value);
              youtubeTimer.current = debouncedMutation({
                slide,
                timerRef: youtubeTimer,
                mutation: updateYoutube,
                values: { youtube: value },
              });
            }}
          />
        </Form.Field>
        <Form.Field>
          <label>Soundcloud: </label>
          <Input
            placeholder="Soundcloud URL"
            icon="linkify"
            iconPosition="left"
            value={soundcloud}
            onChange={(e, { value }) => {
              setSoundcloud(value);
              soundcloudTimer.current = debouncedMutation({
                slide,
                timerRef: soundcloudTimer,
                mutation: updateSoundcloud,
                values: { soundcloud: value },
              });
            }}
          />
        </Form.Field>
      </Segment>
    </Form.Field>
  );
};

Media.propTypes = {
  slide: PropTypes.string.isRequired,
};

export default Media;
