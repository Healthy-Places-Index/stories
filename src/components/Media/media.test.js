import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';

import Media, { GET_SLIDE_MEDIA } from './index';

const mocks = [
  {
    request: {
      query: GET_SLIDE_MEDIA,
      variables: {
        slide: '1',
      },
    },
    result: {
      data: {
        Slide: {
          id: '1',
          youtube: 'http://youtube.com',
          soundcloud: 'http://soundcloud.com',
        },
      },
    },
  },
];

jest.mock('semantic-ui-react', () => {
  const Segment = props => <div {...props} className="segment" />;
  const Field = props => <div {...props} className="field" />;
  const Input = props => <input {...props} />;
  return { Segment, Form: { Field }, Input };
});

describe('Edit Media', () => {
  it('matches snapshot', async () => {
    const component = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Media slide="1" />
      </MockedProvider>
    );

    await act(() => new Promise(resolve => setTimeout(resolve, 0)));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
