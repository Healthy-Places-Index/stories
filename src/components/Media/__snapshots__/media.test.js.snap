// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Media matches snapshot 1`] = `
<div
  className="field"
>
  <label>
    Media
  </label>
  <div
    className="segment"
    style={
      Object {
        "marginTop": 0,
      }
    }
  >
    <div
      className="field"
    >
      <label>
        Youtube: 
      </label>
      <input
        icon="linkify"
        iconPosition="left"
        onChange={[Function]}
        placeholder="Youtube URL"
        value="http://youtube.com"
      />
    </div>
    <div
      className="field"
    >
      <label>
        Soundcloud: 
      </label>
      <input
        icon="linkify"
        iconPosition="left"
        onChange={[Function]}
        placeholder="Soundcloud URL"
        value="http://soundcloud.com"
      />
    </div>
  </div>
</div>
`;
